File: pkg/bus/subjects.go
package bus

// Subject layout:
//
//   {prefix}.events.{event_type}                  // broadcast
//   {prefix}.invoke.{target}                      // request entry
//   {prefix}.{tenant}.events.{event_type}         // tenant-scoped broadcast
//   {prefix}.{tenant}.invoke.{target}             // tenant-scoped request
//
// _INBOX subjects are allocated by the transport (NATS) and embedded in
// envelope.ReplyTo; the bus does not construct them.

func (b *defaultBus) eventSubject(eventType, tenant string) string {
	if tenant != "" {
		return b.opts.SubjectPrefix + "." + tenant + ".events." + eventType
	}
	return b.opts.SubjectPrefix + ".events." + eventType
}

func (b *defaultBus) invokeSubject(target, tenant string) string {
	if tenant != "" {
		return b.opts.SubjectPrefix + "." + tenant + ".invoke." + target
	}
	return b.opts.SubjectPrefix + ".invoke." + target
}

// resolveTenant picks the per-message tenant ID, falling back to the
// bus-level Tenant option.
func (b *defaultBus) resolveTenant(envelopeTenant string) string {
	if envelopeTenant != "" {
		return envelopeTenant
	}
	return b.opts.Tenant
}

File: sdk/python/src/agentflowbus/bus/subjects.py
"""Subject layout helpers. Mirrors pkg/bus/subjects.go.

Subject schema::

  {prefix}.events.{event_type}                   # broadcast
  {prefix}.invoke.{target}                       # request entry
  {prefix}.{tenant}.events.{event_type}          # tenant-scoped broadcast
  {prefix}.{tenant}.invoke.{target}              # tenant-scoped request

_INBOX subjects are allocated by the transport driver and embedded in
``Envelope.reply_to``; the bus does not construct them.
"""

DEFAULT_SUBJECT_PREFIX = "acp.v1"


def event_subject(prefix: str, event_type: str, tenant: str = "") -> str:
    if tenant:
        return f"{prefix}.{tenant}.events.{event_type}"
    return f"{prefix}.events.{event_type}"


def invoke_subject(prefix: str, target: str, tenant: str = "") -> str:
    if tenant:
        return f"{prefix}.{tenant}.invoke.{target}"
    return f"{prefix}.invoke.{target}"
